tools/system_libs.py

```diff
@@ -293,19 +293,40 @@ def create_compiler_rt(libname):
     shared.Building.emar('cr', in_temp(libname), o_s)
     return in_temp(libname)
 
-  def dlmalloc_name():
-    ret = 'dlmalloc'
+  # decides which malloc to use, and returns the source for malloc and the full library name
+  def malloc_decision():
+    if shared.Settings.MALLOC == 'dlmalloc':
+      base = 'dlmalloc'
+    elif shared.Settings.MALLOC == 'emmalloc':
+      base = 'emmalloc'
+    else:
+      raise Exception('malloc must be one of "emmalloc", "dlmalloc", see settings.js')
+    # only dlmalloc supports most modes
+    extra = ''
     if shared.Settings.USE_PTHREADS:
-      ret += '_threadsafe'
+      extra += '_threadsafe'
+      base = 'dlmalloc'
```

Comment:
This kind of validation ends up silently ignoring `-s MALLOC=emmalloc` directive when `-s USE_PTHREADS=1` and similar are set. Perhaps emcc.py could instead validate that `-s MALLOC=emmalloc` and `-s USE_PTHREADS=1` are mutually exclusive and so forth, and then here we could just pass-through `base = shared.Settings.MALLOC`?

Reply:
Good point, I'll make it show a clear error in those cases, instead of silently "fixing" things.